File: www/ChatBot.php
<?php
//////////////////////////////////////////////////////////////////////////
// ChatBot
// -------------------
// begin      : 2020-08-24
// copyright  : Ben Kietzman
// email      : ben@kietzman.org
//////////////////////////////////////////////////////////////////////////
/*! \file ChatBot.php
* \brief ChatBot Class
*
* Provides chat bot functionality.
*/

//! ChatBot Class
/*!
* Provides chat bot functionality.
*/
class ChatBot
{
  // {{{ variables
  protected $m_bConnected;
  protected $m_bDebug;
  protected $m_bQuit;
  protected $m_bSecure;
  protected $m_formArguments;
  protected $m_handle;
  protected $m_messages;
  protected $m_rooms;
  protected $m_strBuffer;
  protected $m_streamContext;
  protected $m_strFormAction;
  protected $m_strFormFooter;
  protected $m_strFormHeader;
  protected $m_strUser;
  protected $m_pMessage;
  // }}}
  // {{{ __construct()
  public function __construct()
  {
    $this->m_bConnected = false;
    $this->m_bDebug = false;
    $this->m_bQuit = false;
    $this->m_bSecure = true;
    $this->m_formArguments = [];
    $this->m_handle = false;
    $this->m_pMessage = null;
    $this->m_messages = [];
    $this->m_rooms = [];
    $this->m_strBuffer = [null, null];
    $context = array();
    $context['ssl'] = array();
    $context['ssl']['verify_peer'] = false;
    $this->m_streamContext = stream_context_create($context);
    unset($context);
  }
  // }}}
  // {{{ __destruct()
  public function __destruct()
  {
    if ($this->m_bConnected)
    {
      $this->disconnect();
    }
    unset($this->m_formArguments);
    unset($this->m_messages);
    unset($this->m_rooms);
    unset($this->m_strBuffer);
  }
  // }}}
  // {{{ analyze()
  private function analyze(&$message)
  {
    $strError = null;

    if ($this->m_pMessage != null)
    {
      call_user_func($this->m_pMessage, $message);
    }
    // {{{ Function
    if (is_array($message) && isset($message['Function']))
    {
      $bStatus = false;
      $strFunction = $message['Function'];
      $strStatus = null;
      if (isset($message['Status']))
      {
        $bStatus = true;
        $strStatus = $message['Status'];
      }
      // {{{ connect
      if ($strFunction == 'connect')
      {
        if ($strStatus == 'okay')
        {
          if (isset($message['Response']) && is_array($message['Response']) && isset($message['Response']['Rooms']) && is_array($message['Response']['Rooms']))
          {
            foreach ($message['Response']['Rooms'] as $key => $value)
            {
              $this->m_rooms[] = $key;
            }
            $this->m_rooms = array_unique($this->m_rooms, SORT_STRING);
          }
          $this->m_bConnected = true;
        }
        else if ($bStatus)
        {
          $this->disconnect();
        }
      }
      // }}}
      // {{{ disconnect
      else if ($strFunction == 'disconnect')
      {
        $this->close($strError);
      }
      // }}}
      // {{{ join
      else if ($strFunction == 'join')
      {
        if ($strStatus == 'okay')
        {
          if (isset($message['Request']) && is_array($message['Request']) && isset($message['Request']['Room']) && $message['Request']['Room'] != '')
          {
            $this->m_rooms[] = $message['Request']['Room'];
            $this->m_rooms = array_unique($this->m_rooms, SORT_STRING);
          }
        }
      }
      // }}}
      // {{{ part
      else if ($strFunction == 'part')
      {
        if ($strStatus == 'okay')
        {
          if (isset($message['Request']) && is_array($message['Request']) && isset($message['Request']['Room']) && $message['Request']['Room'] != '')
          {
            $rooms = [];
            $nSize = sizeof($this->m_rooms);
            for ($i = 0; $i < $nSize; $i++)
            {
              if ($this->m_rooms[$i] != $message['Request']['Room'])
              {
                $rooms[] = $this->m_rooms[$i];
              }
            }
            unset($this->m_rooms);
            $this->m_rooms = $rooms;
            unset($rooms);
            if ($this->m_bQuit)
            {
              if (sizeof($this->m_rooms) > 0)
              {
                $this->part($this->m_rooms[0]);
              }
              else
              {
                $this->disconnect();
              }
            }
          }
        }
      }
      // }}}
    }
    // }}}
  }
  // }}}
  // {{{ close()
  public function close(&$strError)
  {
    $bResult = false;

    $this->m_bConnected = false;
    if ($this->m_handle === false || stream_socket_shutdown($this->m_handle, STREAM_SHUT_RDWR) !== false)
    {
      $bResult = true;
    }
    else
    {
      $strError = 'stream_socket_shutdown() Failed to shutdown.';
    }
    $this->m_handle = false;
    unset($this->m_messages);

    return $bResult;
  }
  // }}}
  // {{{ connect()
  public function connect($strServer, $strPort, $strUser, $strPassword, $strName, &$strError)
  {
    $bResult = false;
    $nErrorNo = null;

    $this->m_strUser = $strUser;
    if ($this->m_handle = stream_socket_client((($this->m_bSecure)?'tls://':'').$strServer.':'.$strPort, $nErrorNo, $strError, 10, STREAM_CLIENT_CONNECT, $this->m_streamContext))
    {
      $bResult = true;
      $message = [];
      $message['Section'] = 'chat';
      $message['Function'] = 'connect';
      $message['User'] = $strUser;
      $message['Password'] = $strPassword;
      $message['Request'] = [];
      $message['Request']['Name'] = $strName;
      $message['wsRequestID'] = -1;
      $this->putMessage($message);
    }
    else
    {
      $strError = 'stream_socket_client() '.$strError;
    }

    return $bResult;
  }
  // }}}
  // {{{ disconnect()
  public function disconnect()
  {
    $message = [];

    $message['Function'] = 'disconnect';
    $message['Request'] = [];
    $this->putMessage($message);
    unset($message);
  }
  // }}}
  // {{{ form()
  public function form($strIdent, $strForm, $arguments = [], $strHeader = '', $strFooter = '')
  {
    $ssForm = '<form onsubmit="fetch(\''.$this->m_strFormAction.'\', {method: \'POST\', cache: \'no-cache\', body: new URLSearchParams(new FormData(this))}).then(response =&gt; {}); return false;">';
    $ssForm .= '<input type="hidden" name="botIdent" value="'.$strIdent.'">';
    foreach ($this->m_formArguments as $key => $value)
    {
      $ssForm .= '<input type="hidden" name="'.$key.'" value="'.$value.'">';
    }
    foreach ($arguments as $key => $value)
    {
      $ssForm .= '<input type="hidden" name="'.$key.'" value="'.$value.'">';
    }
    $ssForm .= '<div style="border-style: solid; border-width: 1px; border-color: #4e5964; border-radius: 10px; background: #2c3742; padding: 10px; color: white;">';
    if ($this->m_strFormHeader != '')
    {
      $ssForm .= $this->m_strFormHeader;
    }
    if ($strHeader != '')
    {
      $ssForm .= $strHeader;
    }
    $ssForm .= $strForm;
    if ($strFooter != '')
    {
      $ssForm .= $strFooter;
    }
    if ($this->m_strFormFooter != '')
    {
      $ssForm .= $this->m_strFormFooter;
    }
    $ssForm .= '</div>';
    $ssForm .= '</form>';

    return $ssForm;
  }
  // }}}
  // {{{ ident()
  public function ident($strUser)
  {
    return md5($strUser.','.microtime(true));
  }
  // }}}
  // {{{ join()
  public function join($strRoom)
  {
    $message = [];

    $message['Function'] = 'join';
    $message['Request'] = [];
    $message['Request']['Room'] = $strRoom;
    $this->putMessage($message);
    unset($message);
  }
  // }}}
  // {{{ message()
  public function message($strTarget, $strMessage, $strText = '')
  {
    $message = [];

    $message['Function'] = 'message';
    $message['Request'] = [];
    $message['Request']['Target'] = $strTarget;
    $message['Request']['Message'] = $strMessage;
    if ($strText != '')
    {
      $message['Request']['Text'] = $strText;
    }
    $this->putMessage($message);
  }
  // }}}
  // {{{ part()
  public function part($strRoom)
  {
    $message = [];

    $message['Function'] = 'part';
    $message['Request'] = [];
    $message['Request']['Room'] = $strRoom;
    $this->putMessage($message);
    unset($message);
  }
  // }}}
  // {{{ ping()
  public function ping()
  {
    $message = [];

    $message['Function'] = 'ping';
    $message['Request'] = [];
    $this->putMessage($message);
    unset($message);
  }
  // }}}
  // {{{ process()
  public function process($nTimeout, &$strError)
  {
    $bResult = false;

    if ($this->m_handle !== false)
    {
      $readfds = [$this->m_handle];
      $writefds = [];
      if ($this->m_strBuffer[1] != '')
      {
        $writefds[] = $this->m_handle;
      }
      else if (sizeof($this->m_messages) > 0)
      {
        $debug = $this->m_messages[0];
        if ($this->m_bConnected || (is_array($debug) && isset($debug['Function']) && $debug['Function'] == 'connect'))
        {
          $writefds[] = $this->m_handle;
          if ($this->m_bDebug)
          {
            if (is_array($debug) && isset($debug['Password']))
            {
              $debug['Password'] = '******';
            }
            echo 'WRITE:  '.json_encode($debug)."\n";
          }
          $this->m_strBuffer[1] .= json_encode($this->m_messages[0])."\n";
          $messages = [];
          $nSize = sizeof($this->m_messages);
          for ($i = 1; $i < $nSize; $i++)
          {
            $messages[] = $this->m_messages[$i];
          }
          unset($this->m_messages);
          $this->m_messages = $messages;
          unset($messages);
        }
        unset($debug);
      }
      $errorfds = null;
      if (($nReturn = stream_select($readfds, $writefds, $errorfds, 0, ($nTimeout * 1000))) > 0)
      {
        if (in_array($this->m_handle, $writefds))
        {
          if (($nReturn = fwrite($this->m_handle, $this->m_strBuffer[1])) !== false)
          {
            $bResult = true;
            $this->m_strBuffer[1] = substr($this->m_strBuffer[1], $nReturn, (strlen($this->m_strBuffer[1]) - $nReturn));
          }
          else
          {
            $strError = 'fwrite() Failed to write.';
          }
        }
        if (in_array($this->m_handle, $readfds))
        {
          if (($strBuffer = fread($this->m_handle, 65536)) !== false)
          {
            if ($strBuffer != '')
            {
              $bResult = true;
              $this->m_strBuffer[0] .= $strBuffer;
              while (($unPosition = strpos($this->m_strBuffer[0], "\n")) !== false)
              {
                $message = json_decode(substr($this->m_strBuffer[0], 0, $unPosition), true);
                $this->m_strBuffer[0] = substr($this->m_strBuffer[0], ($unPosition + 1), (strlen($this->m_strBuffer[0]) - ($unPosition + 1)));
                if ($this->m_bDebug)
                {
                  $debug = $message;
                  if (is_array($debug) && isset($debug['Password']))
                  {
                    $debug['Password'] = '******';
                  }
                  echo 'READ:  '.json_encode($debug)."\n";
                  unset($debug);
                }
                $this->analyze($message);
                unset($message);
              }
            }
          }
          else
          {
            $strError = 'fread() Failed to read.';
          }
        }
      }
      else if ($nReturn === false)
      {
        $strError = 'stream_select() Failed to select.';
      }
      else
      {
        $bResult = true;
      }
      if (!$bResult)
      {
        $this->close($strError);
      }
    }
    else
    {
      $strError = 'Please provide the stream resource.';
      $this->close($strError);
    }
    return $bResult;
  }
  // }}}
  // {{{ putMessage()
  public function putMessage(&$message)
  {
    $this->m_messages[] = $message;
  }
  // }}}
  // {{{ quit()
  public function quit()
  {
    $this->m_bQuit = true;
    if (sizeof($this->m_rooms) > 0)
    {
      $this->part($this->m_rooms[0]);
    }
    else
    {
      $this->disconnect();
    }
  }
  // }}}
  // {{{ rooms()
  public function rooms()
  {
    return $this->m_rooms;
  }
  // }}}
  // {{{ secure()
  public function secure($bSecure = true)
  {
    $this->m_bSecure = $bSecure;
  }
  // }}}
  // {{{ setDebug()
  public function setDebug($bDebug)
  {
    $this->m_bDebug = $bDebug;
  }
  // }}}
  // {{{ setForm()
  public function setForm($strAction, $arguments = [], $strHeader = '', $strFooter = '')
  {
    $this->m_strFormAction = $strAction;
    if (sizeof($arguments) > 0)
    {
      $this->m_formArguments = $arguments;
    }
    if ($strHeader != '')
    {
      $this->m_strFormHeader = $strHeader;
    }
    if ($strFooter != '')
    {
      $this->m_strFormFooter = $strFooter;
    }
  }
  // }}}
  // {{{ setFormArguments()
  public function setFormArguments($arguments)
  {
    $this->m_formArguments = $arguments;
  } 
  // }}}
  // {{{ setFormFooter()
  public function setFormFooter($strFooter)
  {
    $this->m_strFormFooter = $strFooter;
  } 
  // }}}
  // {{{ setFormHeader()
  public function setFormHeader($strHeader)
  {
    $this->m_strFormHeader = $strHeader;
  }
  // }}}
  // {{{ setMessage()
  public function setMessage($pMessage)
  {
    $this->m_pMessage = $pMessage;
  }
  // }}}
}
?>

File: ChatBot.cpp
// vim: syntax=cpp
// vim600: fdm=marker
/* -*- c++ -*- */
///////////////////////////////////////////
// Chat Bot
// -------------------------------------
// file       : ChatBot.cpp
// author     : Ben Kietzman
// begin      : 2018-01-12
// copyright  : kietzman.org
// email      : ben@kietzman.org
///////////////////////////////////////////

/**************************************************************************
*                                                                         *
*   This program is free software; you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation; either version 2 of the License, or     *
*   (at your option) any later version.                                   *
*                                                                         *
**************************************************************************/

/*! \file ChatBot.cpp
* \brief ChatBot Class
*/
// {{{ includes
#include "ChatBot"
// }}}
extern "C++"
{
  namespace common
  {
    // {{{ ChatBot()
    ChatBot::ChatBot()
    {
      string strError;

      m_bConnected = false;
      m_bDebug = false;
      m_bQuit = false;
      m_bSecure = true;
      m_ctx = NULL;
      m_fdSocket = -1;
      m_pMessage = NULL;
      m_pUtility = new Utility(strError);
      m_ssl = NULL;
    }
    // }}}
    // {{{ ~ChatBot()
    ChatBot::~ChatBot()
    {
      if (m_bConnected)
      {
        disconnect();
      }
      m_formArguments.clear();
      delete m_pUtility;
    }
    // }}}
    // {{{ analyze()
    void ChatBot::analyze(Json *ptMessage)
    {
      if (m_pMessage != NULL)
      {
        (*m_pMessage)(ptMessage);
      }
      // {{{ Function
      if (ptMessage->m.find("Function") != ptMessage->m.end())
      {
        bool bStatus = false;
        string strFunction = ptMessage->m["Function"]->v, strStatus;
        if (ptMessage->m.find("Status") != ptMessage->m.end())
        {
          bStatus = true;
          strStatus = ptMessage->m["Status"]->v;
        }
        // {{{ connect
        if (strFunction == "connect")
        {
          if (strStatus == "okay")
          {
            if (ptMessage->m.find("Response") != ptMessage->m.end() && ptMessage->m["Response"]->m.find("Rooms") != ptMessage->m["Response"]->m.end())
            {
              for (map<string, Json *>::iterator i = ptMessage->m["Response"]->m["Rooms"]->m.begin(); i != ptMessage->m["Response"]->m["Rooms"]->m.end(); i++)
              {
                m_rooms.push_back(i->first);
              }
              m_rooms.sort();
              m_rooms.unique();
            }
            m_bConnected = true;
          }
          else if (bStatus)
          {
            disconnect();
          }
        }
        // }}}
        // {{{ disconnect
        else if (strFunction == "disconnect")
        {
          close();
        }
        // }}}
        // {{{ join
        else if (strFunction == "join")
        {
          if (strStatus == "okay")
          {
            if (ptMessage->m.find("Request") != ptMessage->m.end() && ptMessage->m["Request"]->m.find("Room") != ptMessage->m["Request"]->m.end() && !ptMessage->m["Request"]->m["Room"]->v.empty())
            {
              m_rooms.push_back(ptMessage->m["Request"]->m["Room"]->v);
              m_rooms.sort();
              m_rooms.unique();
            }
          }
        }
        // }}}
        // {{{ part
        else if (strFunction == "part")
        {
          if (strStatus == "okay")
          {
            if (ptMessage->m.find("Request") != ptMessage->m.end() && ptMessage->m["Request"]->m.find("Room") != ptMessage->m["Request"]->m.end() && !ptMessage->m["Request"]->m["Room"]->v.empty())
            {
              list<string>::iterator removeIter = m_rooms.end();
              for (list<string>::iterator i = m_rooms.begin(); removeIter == m_rooms.end() && i != m_rooms.end(); i++)
              {
                if ((*i) == ptMessage->m["Request"]->m["Room"]->v)
                {
                  removeIter = i;
                }
              }
              if (removeIter != m_rooms.end())
              {
                m_rooms.erase(removeIter);
              }
              if (m_bQuit)
              {
                if (!m_rooms.empty())
                {
                  part(m_rooms.front());
                }
                else
                {
                  disconnect();
                }
              }
            }
          }
        }
        // }}}
      }
      // }}}
    }
    // }}}
    // {{{ close()
    bool ChatBot::close(string &strError)
    {
      bool bResult = false;
      stringstream ssError;

      m_bConnected = false;
      if (m_ssl != NULL)
      {
        SSL_shutdown(m_ssl);
        SSL_free(m_ssl);
        m_ssl = NULL;
      }
      if (m_ctx != NULL)
      {
        SSL_CTX_free(m_ctx);
        m_ctx = NULL;
      }
      if (m_fdSocket == -1)
      {
        if (::close(m_fdSocket) == 0)
        {
          bResult = true;
        }
        else
        {
          ssError.str("");
          ssError << "close(" << errno << ") " << strerror(errno);
          strError = ssError.str();
        }
        m_fdSocket = -1;
      }
      while (!m_messages.empty())
      {
        delete m_messages.front();
        m_messages.pop_front();
      }

      return bResult;
    }
    void ChatBot::close()
    {
      string strError;

      close(strError);
    }
    // }}}
    // {{{ connect()
    bool ChatBot::connect(const string strServer, const string strPort, const string strUser, const string strPassword, const string strName, string &strError)
    {
      bool bResult = false;
      addrinfo hints, *result;
      int fdSocket, nReturn;
      stringstream ssError;

      m_strUser = strUser;
      memset(&hints, 0, sizeof(addrinfo));
      hints.ai_family = AF_UNSPEC;
      hints.ai_socktype = SOCK_STREAM;
      if (m_bSecure)
      {
        SSL_METHOD *method = (SSL_METHOD *)SSLv23_client_method();
        if ((m_ctx = SSL_CTX_new(method)) == NULL)
        {
          m_bSecure = false;
        }
      }
      if ((nReturn = getaddrinfo(strServer.c_str(), strPort.c_str(), &hints, &result)) == 0)
      {
        bool bSocket = false;
        addrinfo *rp;
        for (rp = result; m_fdSocket == -1 && rp != NULL; rp = rp->ai_next)
        {
          bSocket = false;
          if ((fdSocket = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol)) >= 0)
          {
            bSocket = true;
            if (::connect(fdSocket, rp->ai_addr, rp->ai_addrlen) == 0)
            {
              if (!m_bSecure || (m_ssl = m_pUtility->sslConnect(m_ctx, fdSocket, strError)) != NULL)
              {
                m_fdSocket = fdSocket;
              }
              else
              {
                ::close(fdSocket);
              }
            }
            else
            {
              ::close(fdSocket);
            }
          }
        }
        freeaddrinfo(result);
        if (m_fdSocket != -1)
        {
          Json *ptMessage = new Json;
          bResult = true;
          ptMessage->insert("Section", "chat");
          ptMessage->insert("Function", "connect");
          ptMessage->insert("User", strUser);
          ptMessage->insert("Password", strPassword);
          ptMessage->m["Request"] = new Json;
          ptMessage->m["Request"]->insert("Name", strName);
          ptMessage->insert("wsRequestID", "-1", 'n');
          putMessage(ptMessage);
        }
        else
        {
          ssError.str("");
          ssError << ((bSocket)?"connect":"socket") << "(" << errno << ") " << strerror(errno);
          strError = ssError.str();
        }
      }
      else
      {
        ssError.str("");
        ssError << "getaddrinfo(" << nReturn << ") " << gai_strerror(nReturn);
        strError = ssError.str();
      }

      return bResult;
    }
    // }}}
    // {{{ disconnect()
    void ChatBot::disconnect()
    {
      putMessage("disconnect");
    }
    // }}}
    // {{{ form()
    string ChatBot::form(const string strIdent, const string strForm, const string strHeader, const string strFooter)
    {
      map<string, string> arguments;

      return form(strIdent, strForm, arguments, strHeader, strFooter);
    }
    string ChatBot::form(const string strIdent, const string strForm, map<string, string> arguments, const string strHeader, const string strFooter)
    {
      stringstream ssForm;

      ssForm << "<form onsubmit=\"fetch('" << m_strFormAction << "', {method: 'POST', cache: 'no-cache', body: new URLSearchParams(new FormData(this))}).then(response =&gt; {}); return false;\">";
      ssForm << "<input type=\"hidden\" name=\"botIdent\" value=\"" << strIdent << "\">";
      for (map<string, string>::iterator i = m_formArguments.begin(); i != m_formArguments.end(); i++)
      {
        ssForm << "<input type=\"hidden\" name=\"" << i->first << "\" value=\"" << i->second << "\">";
      }
      for (map<string, string>::iterator i = arguments.begin(); i != arguments.end(); i++)
      {
        ssForm << "<input type=\"hidden\" name=\"" << i->first << "\" value=\"" << i->second << "\">";
      }
      ssForm << "<div style=\"border-style: solid; border-width: 1px; border-color: #4e5964; border-radius: 10px; background: #2c3742; padding: 10px; color: white;\">";
      if (!m_strFormHeader.empty())
      {
        ssForm << m_strFormHeader;
      }
      if (!strHeader.empty())
      {
        ssForm << strHeader;
      }
      ssForm << strForm;
      if (!strFooter.empty())
      {
        ssForm << strFooter;
      }
      if (!m_strFormFooter.empty())
      {
        ssForm << m_strFormFooter;
      }
      ssForm << "</div>";
      ssForm << "</form>";

      return ssForm.str();
    }
    // }}}
    // {{{ ident()
    string ChatBot::ident(const string strUser)
    {
      char md5string[33];
      MD5_CTX context;
      string strIdent;
      stringstream ssIdent;
      timespec start;
      unsigned char digest[16];

      clock_gettime(CLOCK_REALTIME, &start);
      ssIdent << strUser << "," << start.tv_sec << "," << start.tv_nsec;
      MD5_Init(&context);
      MD5_Update(&context, ssIdent.str().c_str(), ssIdent.str().size());
      MD5_Final(digest, &context);
      for (int i = 0; i < 16; i++)
      {
        sprintf(&md5string[i*2], "%02x", (unsigned int)digest[i]);
      }
      strIdent = md5string;

      return strIdent;
    }
    // }}}
    // {{{ join()
    void ChatBot::join(const string strRoom)
    {
      Json *ptMessage = new Json;

      ptMessage->insert("Room", strRoom);
      putMessage("join", ptMessage);
    }
    // }}}
    // {{{ message()
    void ChatBot::message(const string strTarget, const string strMessage, const string strText)
    {
      Json *ptMessage = new Json;

      ptMessage->insert("Target", strTarget);
      ptMessage->insert("Message", strMessage);
      if (!strText.empty())
      {
        ptMessage->insert("Text", strText);
      }
      putMessage("message", ptMessage);
    }
    // }}}
    // {{{ part()
    void ChatBot::part(const string strRoom)
    {
      Json *ptMessage = new Json;

      ptMessage->insert("Room", strRoom);
      putMessage("part", ptMessage);
    }
    // }}}
    // {{{ ping()
    void ChatBot::ping()
    {
      putMessage("ping");
    }
    // }}}
    // {{{ process()
    bool ChatBot::process(int nTimeout, string &strError)
    {
      bool bResult = false;
      stringstream ssError;

      if (m_fdSocket != -1)
      {
        int nReturn;
        pollfd fds[1];
        size_t unPosition;
        fds[0].fd = m_fdSocket;
        fds[0].events = POLLIN;
        if (!m_strBuffer[1].empty())
        {
          fds[0].events |= POLLOUT;
        }
        else if (!m_messages.empty())
        {
          Json *ptDebug = new Json(m_messages.front());
          if (m_bConnected || (ptDebug->m.find("Function") != ptDebug->m.end() && ptDebug->m["Function"]->v == "connect"))
          {
            stringstream ssJson;
            fds[0].events |= POLLOUT;
            if (m_bDebug)
            {
              if (ptDebug->m.find("Password") != ptDebug->m.end())
              {
                ptDebug->insert("Password", "******");
              }
              cout << "WRITE:  " << ptDebug << endl;
            }
            ssJson << m_messages.front();
            delete m_messages.front();
            m_messages.pop_front();
            m_strBuffer[1] += ssJson.str() + "\n";
          }
          delete ptDebug;
        }
        if ((nReturn = poll(fds, 1, nTimeout)) > 0)
        {
          if (fds[0].revents & POLLIN)
          {
            if ((m_ssl != NULL && m_pUtility->sslRead(m_ssl, m_strBuffer[0], nReturn)) || (m_ssl == NULL && m_pUtility->fdRead(m_fdSocket, m_strBuffer[0], nReturn)))
            {
              bResult = true;
              while ((unPosition = m_strBuffer[0].find("\n")) != string::npos)
              {
                Json *ptMessage = new Json(m_strBuffer[0].substr(0, unPosition));
                m_strBuffer[0].erase(0, (unPosition + 1));
                if (m_bDebug)
                {
                  Json *ptDebug = new Json(ptMessage);
                  if (ptDebug->m.find("Password") != ptDebug->m.end())
                  {
                    ptDebug->insert("Password", "******");
                  }
                  cout << "READ:  " << ptDebug << endl;
                  delete ptDebug;
                }
                analyze(ptMessage);
                delete ptMessage;
              }
            }
            else
            {
              ssError.str("");
              if (m_ssl != NULL)
              {
                ssError << "Utility::sslRead(" << nReturn << ") " << m_pUtility->sslstrerror(m_ssl, nReturn);
              }
              else
              {
                ssError << "Utility::fdRead(" << errno << ") " << strerror(errno);
              }
              strError = ssError.str();
            }
          }
          if (fds[0].revents & POLLOUT)
          {
            if ((m_ssl != NULL && m_pUtility->sslWrite(m_ssl, m_strBuffer[1], nReturn)) || (m_ssl == NULL && m_pUtility->fdWrite(m_fdSocket, m_strBuffer[1], nReturn)))
            {
              bResult = true;
            }
            else
            {
              ssError.str("");
              if (m_ssl != NULL)
              {
                ssError << "Utility::sslWrite(" << nReturn << ") " << m_pUtility->sslstrerror(m_ssl, nReturn);
              }
              else
              {
                ssError << "Utility::fdWrite(" << errno << ") " << strerror(errno);
              }
              strError = ssError.str();
            }
          }
        }
        else if (nReturn == 0 || errno == 4)
        {
          bResult = true;
        }
        else
        {
          ssError.str("");
          ssError << "poll(" << errno << ") " << strerror(errno);
          strError = ssError.str();
        }
        if (!bResult)
        {
          close();
        }
      }
      else
      {
        strError = "Please provide the socket file descriptor.";
        close();
      }

      return bResult;
    }
    // }}}
    // {{{ putMessage()
    void ChatBot::putMessage(Json *ptMessage)
    {
      m_messages.push_back(ptMessage);
    }
    void ChatBot::putMessage(const string strFunction, Json *ptMessage)
    {
      Json *ptRequest = new Json;
 
      ptRequest->insert("Function", strFunction);
      ptRequest->m["Request"] = ((ptMessage != NULL)?ptMessage:new Json);
      m_messages.push_back(ptRequest);
    }
    // }}}
    // {{{ quit()
    void ChatBot::quit()
    {
      m_bQuit = true;
      if (!m_rooms.empty())
      {
        part(m_rooms.front());
      }
      else
      {
        disconnect();
      }
    }
    // }}}
    // {{{ rooms()
    list<string> ChatBot::rooms()
    {
      return m_rooms;
    }
    // }}}
    // {{{ setDebug()
    void ChatBot::setDebug(const bool bDebug)
    {
      m_bDebug = bDebug;
    }
    // }}}
    // {{{ setForm()
    void ChatBot::setForm(const string strAction, const string strHeader, const string strFooter)
    {
      map<string, string> arguments;

      setForm(strAction, arguments, strHeader, strFooter);
    }
    void ChatBot::setForm(const string strAction, map<string, string> arguments, const string strHeader, const string strFooter)
    {
      m_strFormAction = strAction;
      if (!arguments.empty())
      {
        m_formArguments = arguments;
      }
      if (!strHeader.empty())
      {
        m_strFormHeader = strHeader;
      }
      if (!strFooter.empty())
      {
        m_strFormFooter = strFooter;
      }
    }
    // }}}
    // {{{ setFormArguments()
    void ChatBot::setFormArguments(map<string, string> arguments)
    {
      m_formArguments = arguments;
    }
    // }}}
    // {{{ setFormFooter()
    void ChatBot::setFormFooter(const string strFooter)
    {
      m_strFormFooter = strFooter;
    }
    // }}}
    // {{{ setFormHeader()
    void ChatBot::setFormHeader(const string strHeader)
    {
      m_strFormHeader = strHeader;
    }
    // }}}
    // {{{ setMessage()
    void ChatBot::setMessage(void (*pMessage)(Json *))
    {
      m_pMessage = pMessage;
    }
    // }}}
  }
}
